### PR TITLE
OCPBUGS-9998: Ambient credentials support

### PR DESCRIPTION
**Pull request by swghosh**
- Add `--issuer-ambient-credentials` as supported cert-manager controller arg
- Service account bound token using projected volume "/var/run/secrets/openshift/serviceaccount/token"

Signed-off-by: Swarup Ghosh <swghosh@redhat.com>

### CODE REVIEW

**Comment by swghosh** [comment]
/hold
till we figure out if mounting the cloud credentials secret to the deployment pod itself is required or just bound-sa-token would suffice

**Comment by swghosh** [comment]
/hold
the secrets informer implemented in this PR has an empty cache (probably, a bug or something we're doing wrong during initialization)

File: pkg/controller/deployment/credentials_request.go
Comment (stlaz) [inline pkg/controller/deployment/credentials_request.go:96]: don't you have operator configuration that you can read this from?

Comment (stlaz) [inline pkg/controller/deployment/credentials_request.go:96]:
```go
secretName := os.Getenv("CLOUD_SECRET_NAME")
if len(secretName) == 0 {
    return nil
}
```

**Review by stlaz — CHANGES_REQUESTED** [review]

File: pkg/controller/deployment/generic_deployment_controller.go
Comment (stlaz) [inline pkg/controller/deployment/generic_deployment_controller.go:53]: Add the secrets informer to the informer array on line 39.

Not only it will add a wait for its caches to sync upon the controller startup, it's also going to add event handlers to the informer events. Without it, your controller is not going to react to the secret changes.

**Review by stlaz — CHANGES_REQUESTED** [review]

File: pkg/controller/deployment/credentials_request.go
Comment (swghosh) [reply pkg/controller/deployment/credentials_request.go:96]: yup, had this one improvised in another follow-up commit, yet to push it.

File: pkg/controller/deployment/credentials_request.go
Comment (thejasn) [inline pkg/controller/deployment/credentials_request.go:133]: Can you change this to a switch and drop `cloudCredentialConfigs`? I think it would make it more readable and maintainable.

File: pkg/controller/deployment/credentials_request.go
Comment (thejasn) [inline pkg/controller/deployment/credentials_request.go:95]: Can you move this to a separate hook? Since these are more like defaults for an operator running on OCP.

File: pkg/controller/deployment/credentials_request.go
Comment (thejasn) [inline pkg/controller/deployment/credentials_request.go:107]: Duplicate error handling. Merge them.

File: pkg/controller/deployment/deployment_overrides_validation.go
Comment (thejasn) [inline pkg/controller/deployment/deployment_overrides_validation.go:40]: Add note on specific requirement for GCP.

File: pkg/operator/starter.go
Comment (thejasn) [inline pkg/operator/starter.go:61]: We want this informer as well, to enforce resources created in `cert-manager` namespace.

**Review by thejasn — CHANGES_REQUESTED** [review]
Can you update docs as well to include required commands on how to set credentials secret name?

File: pkg/operator/starter.go
Comment (swghosh) [reply pkg/operator/starter.go:61]: `operatorclient.TargetNamespace` already has the same value "cert-manager", as the value was duplicated I ended up removing the former.

File: pkg/controller/deployment/credentials_request.go
Comment (swghosh) [reply pkg/controller/deployment/credentials_request.go]: Yes added a new hook: `withSABoundToken`.

File: pkg/controller/deployment/credentials_request.go
Comment (swghosh) [reply pkg/controller/deployment/credentials_request.go]: Good catch, de-duped.

File: pkg/controller/deployment/credentials_request.go
Comment (swghosh) [reply pkg/controller/deployment/credentials_request.go]: Refactored to use switch case instead.

File: pkg/controller/deployment/credentials_request.go
Comment (swghosh) [reply pkg/controller/deployment/credentials_request.go]: Included in latest commit, thanks!

File: pkg/controller/deployment/deployment_overrides_validation.go
Comment (swghosh) [reply pkg/controller/deployment/deployment_overrides_validation.go:40]: Done

**Comment by swghosh** [comment]
/cc @davemulford 
for px-approved label
/cc @xenolinux 
for docs-approved labe;
/cc @xingxingxia 
for qe-approved, could you please approve this one, feel free to test this one out in our next staging build (post-merge)

File: pkg/controller/deployment/credentials_request.go
Comment (thejasn) [inline pkg/controller/deployment/credentials_request.go:102]: Can you refactor this to so that we don't have multiple `append()`. Create a ref for volume and volumeMount outside the switch. 

Also return error on `default` in the switch, we need to catch use cases of other platforms that are using credentials (STS). WDYT?

Comment (swghosh) [reply pkg/controller/deployment/credentials_request.go]: Cool, used:
```
var volume *corev1.Volume
var volumeMount *corev1.VolumeMount
```
and a single set of appends towards the end of the func.

Regarding, `default` case updated it return an error, this would also help discourage users not on AWS/GCP from using that env var for mounting secrets, thanks!

File: pkg/controller/deployment/credentials_request.go
Comment (thejasn) [inline pkg/controller/deployment/credentials_request.go:35]:
```suggestion
	if deploymentName != certmanagerControllerDeployment {
```

Comment (swghosh) [reply pkg/controller/deployment/credentials_request.go]: Updated in latest.

**Comment by xingxingxia** [comment]
> for qe-approved, could you please approve this one, feel free to test this one out in our next staging build (post-merge)

Got. Though, no matter of pre-merge / post-merge test, I'd like to give /qe-approved after dev peers' /lgtm present.

**Comment by swghosh** [comment]
/jira cc-qa

> Got. Though, no matter of pre-merge / post-merge test, I'd like to give qe-approved after dev peers' lgtm present.

@xingxingxia guess now you can add :)

**Comment by xingxingxia** [comment]
Will do post-merge test.
/label qe-approved

**Pull request merged**
